src/main/scala/org/apache/spark/rdd/OrderedRDD.scala

```diff
@@ -0,0 +1,159 @@
+    }
+
+    rdd match {
+      case _: OrderedRDD[T, K, V] =>
+        rdd.asInstanceOf[OrderedRDD[T, K, V]]
+      case _: EmptyRDD[(K, V)] => new OrderedRDD(rdd, OrderedPartitioner(Array.empty[T]))
+      case _ =>
+        rdd.partitioner match {
+          case Some(p: OrderedPartitioner[T, K]) =>
+            new OrderedRDD(rdd, p)
+          case _ =>
+            val result = reducedRDD.getOrElse(rdd.map(_._1))
+              .mapPartitionsWithIndex(verifySortedness).collect()
+
+            anyFailAllFail[Array, (Int, Boolean, T, T)](result)
+              .map(_.sortBy(_._1))
+              .flatMap(fromPartitionSummaries)
+              .getOrElse(fromShuffle())
+        }
+    }
+  }
+
+  /**
+    * Lifted from org.apache.spark.RangePartitioner.scala in org.apache.spark
+    */
+  def calculateKeyRanges[T](rdd: RDD[T])(implicit ord: Ordering[T], tct: ClassTag[T]): Array[T] = {
+    // This is the sample size we need to have roughly balanced output partitions, capped at 1M.
+    val sampleSize = math.min(20.0 * rdd.partitions.length, 1e6)
+    // Assume the input partitions are roughly balanced and over-sample a little bit.
+    val sampleSizePerPartition = math.ceil(3.0 * sampleSize / rdd.partitions.length).toInt
+    val (numItems, sketched) = RangePartitioner.sketch(rdd, sampleSizePerPartition)
+    if (numItems == 0L) {
+      Array.empty
+    } else {
+      // If a partition contains much more than the average number of items, we re-sample from it
+      // to ensure that enough items are collected from that partition.
+      val fraction = math.min(sampleSize / math.max(numItems, 1L), 1.0)
+      val candidates = ArrayBuffer.empty[(T, Float)]
+      val imbalancedPartitions = mutable.Set.empty[Int]
+      sketched.foreach {
+        case (idx, n, sample) =>
+          if (fraction * n > sampleSizePerPartition) {
+            imbalancedPartitions += idx
+          } else {
+            // The weight is 1 over the sampling probability.
+            val weight = (n.toDouble / sample.length).toFloat
+            for (key <- sample) {
+              candidates += ((key, weight))
+            }
+          }
+      }
+      if (imbalancedPartitions.nonEmpty) {
+        // Re-sample imbalanced partitions with the desired sampling probability.
+        val imbalanced = new PartitionPruningRDD(rdd, imbalancedPartitions.contains)
+        val seed = byteswap32(-rdd.id - 1)
+        val reSampled = imbalanced.sample(withReplacement = false, fraction, seed).collect()
+        val weight = (1.0 / fraction).toFloat
+        candidates ++= reSampled.map(x => (x, weight))
+      }
+      RangePartitioner.determineBounds(candidates, rdd.partitions.length)
+    }
+  }
+
+}
+
+class OrderedRDD[T, K, V](rdd: RDD[(K, V)], p: OrderedPartitioner[T, K])
+  (implicit ev: (K) => T, tOrd: Ordering[T], kOrd: Ordering[K], tct: ClassTag[T], kct: ClassTag[K]) extends RDD[(K, V)](rdd) {
+
+  def orderedPartitioner = p
```

Comment:
This can just be `val orderedPartitioner: OrderedPartitioner[T, K]` in the constructor.